source/module_hsolver/diago_david.h

```diff
@@ -21,10 +21,11 @@ class DiagoDavid
     // return T if T is real type(float, double), 
     // otherwise return the real type of T(complex<float>, complex<double>)
     using Real = typename GetTypeReal<T>::type;
-  
-  public:
 
-    DiagoDavid(const Real* precondition_in,
+    using PreFunc = fvec::Div<T>;
```

Comment:
Is it possible to give PrecondFunc the interface like HPsi or SPsi? 
```cpp
// in precondition lib
using Div = std::function<void(T*, const size_t&, const size_t&)>;
```
Maybe we can give it the form
```cpp
using HPsiFunc = std::function<void(T*, T*, const int, const int)>;
using SPsiFunc = std::function<void(T*, T*, const int, const int)>;
using PrecondFunc = std::function<void(T*, T*, const int, const int)>;
```
and pass the preconditioners to this universal function wrapper.

Reply:
Good point, see the latest change. 
And seems on `HPsiFunc` and `SPsiFunc` can also add "const" like this:
```
using HPsiFunc = std::function<void(T*const, constT*const, const int, const int)>;
```